File: api/routes/flashcards.py
"""Routes for flashcard CRUD operations."""

from __future__ import annotations

import uuid
from typing import Iterable, List

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from api.dependencies import get_current_user, require_csrf
from api.db import models
from api.db.database import get_db
from api.schemas import FlashcardCreate, FlashcardOut, FlashcardUpdate


router = APIRouter(prefix="/flashcards", tags=["flashcards"])


def _ensure_notebook_owned(db: Session, user: models.User, notebook_id: uuid.UUID) -> models.Notebook:
    notebook = db.get(models.Notebook, notebook_id)
    if not notebook or notebook.user_id != user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notebook not found")
    return notebook


def _card_to_schema(card: models.Flashcard) -> FlashcardOut:
    return FlashcardOut(
        id=card.id,
        notebook_id=card.notebook_id,
        question=card.question,
        answer=card.answer,
        meta=card.meta,
        folder_ids=[folder.id for folder in card.folders],
    )


def _get_card(card_id: uuid.UUID, user: models.User, db: Session) -> models.Flashcard:
    card = (
        db.execute(
            select(models.Flashcard)
            .where(models.Flashcard.id == card_id)
            .options(selectinload(models.Flashcard.folders))
        )
        .scalars()
        .first()
    )
    if not card or card.user_id != user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Flashcard not found")
    return card


def _get_folders(
    db: Session,
    user: models.User,
    notebook_id: uuid.UUID,
    folder_ids: Iterable[uuid.UUID],
) -> List[models.FlashcardFolder]:
    ids = list(dict.fromkeys(folder_ids))
    if not ids:
        return []
    folders = (
        db.execute(
            select(models.FlashcardFolder).where(
                models.FlashcardFolder.user_id == user.id,
                models.FlashcardFolder.id.in_(ids),
            )
        )
        .scalars()
        .all()
    )
    found = {folder.id for folder in folders}
    missing = [str(folder_id) for folder_id in ids if folder_id not in found]
    if missing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Flashcard folders not found: {', '.join(missing)}",
        )
    invalid = [str(folder.id) for folder in folders if folder.notebook_id != notebook_id]
    if invalid:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Folders belong to a different notebook: {', '.join(invalid)}",
        )
    return folders


@router.get("", response_model=List[FlashcardOut])
def list_flashcards(
    notebook_id: uuid.UUID | None = Query(default=None),
    user: models.User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> List[FlashcardOut]:
    """List flashcards for the user; optionally filter by notebook id."""
    query = (
        select(models.Flashcard)
        .where(models.Flashcard.user_id == user.id)
        .options(selectinload(models.Flashcard.folders))
        .order_by(models.Flashcard.id)
    )
    if notebook_id is not None:
        query = query.where(models.Flashcard.notebook_id == notebook_id)

    cards = db.execute(query).scalars().unique().all()
    return [_card_to_schema(card) for card in cards]


@router.post(
    "",
    response_model=FlashcardOut,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_csrf)],
)
def create_flashcard(
    payload: FlashcardCreate,
    user: models.User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> FlashcardOut:
    """Create a flashcard under a notebook and optionally attach it to folders."""
    _ensure_notebook_owned(db, user, payload.notebook_id)
    card = models.Flashcard(
        user_id=user.id,
        notebook_id=payload.notebook_id,
        question=payload.question,
        answer=payload.answer,
        meta=payload.meta,
    )
    card.folders = _get_folders(db, user, card.notebook_id, payload.folder_ids)
    db.add(card)
    db.commit()
    db.refresh(card)
    return _card_to_schema(card)


@router.put(
    "/{card_id}",
    response_model=FlashcardOut,
    dependencies=[Depends(require_csrf)],
)
def update_flashcard(
    card_id: uuid.UUID,
    payload: FlashcardUpdate,
    user: models.User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> FlashcardOut:
    """Update flashcard text/metadata and its folder membership."""
    card = _get_card(card_id, user, db)

    if payload.question is not None:
        card.question = payload.question
    if payload.answer is not None:
        card.answer = payload.answer
    if payload.meta is not None:
        card.meta = payload.meta
    if payload.folder_ids is not None:
        card.folders = _get_folders(db, user, card.notebook_id, payload.folder_ids)

    db.add(card)
    db.commit()
    db.refresh(card)
    return _card_to_schema(card)


@router.delete(
    "/{card_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_csrf)],
)
def delete_flashcard(
    card_id: uuid.UUID,
    user: models.User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> Response:
    """Delete a flashcard owned by the user."""
    card = _get_card(card_id, user, db)
    db.delete(card)
    db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: web/src/services/api/flashcards.ts
/**
 * Flashcards API - 闪卡相关接口
 */

import type { Flashcard, FlashcardFolder } from '@/types/flashcards'
import { apiFetch } from './client'
import type { ApiFlashcard, ApiFlashcardFolder } from './types'
import { mapFlashcard, mapFlashcardFolder } from './types'

export const listFlashcardFolders = async (options?: { notebookId?: string }): Promise<FlashcardFolder[]> => {
  const query = options?.notebookId ? `?notebook_id=${encodeURIComponent(options.notebookId)}` : ''
  const data = await apiFetch<ApiFlashcardFolder[]>(`/flashcards/folders${query}`, {
    method: 'GET',
    skipCsrf: true,
  })
  return data.map(mapFlashcardFolder)
}

export const updateFlashcardFolder = async (
  folderId: string,
  payload: { name?: string; description?: string | null; flashcardIds?: string[] },
): Promise<FlashcardFolder> => {
  const body: Record<string, unknown> = {}
  if (payload.name !== undefined) body.name = payload.name
  if (payload.description !== undefined) body.description = payload.description
  if (payload.flashcardIds !== undefined) body.flashcard_ids = payload.flashcardIds

  const data = await apiFetch<ApiFlashcardFolder>(`/flashcards/folders/${folderId}`, { method: 'PUT', body })
  return mapFlashcardFolder(data)
}

export const deleteFlashcardFolder = async (folderId: string): Promise<void> => {
  await apiFetch<void>(`/flashcards/folders/${folderId}`, { method: 'DELETE' })
}

export const listFlashcards = async (options?: { notebookId?: string }): Promise<Flashcard[]> => {
  const query = options?.notebookId ? `?notebook_id=${encodeURIComponent(options.notebookId)}` : ''
  const data = await apiFetch<ApiFlashcard[]>(`/flashcards${query}`, { method: 'GET', skipCsrf: true })
  return data.map(mapFlashcard)
}

export const deleteFlashcard = async (cardId: string): Promise<void> => {
  await apiFetch<void>(`/flashcards/${cardId}`, { method: 'DELETE' })
}

export const updateFlashcard = async (
  cardId: string,
  payload: { question?: string; answer?: string; meta?: Record<string, unknown> | null; folderIds?: string[] },
): Promise<Flashcard> => {
  const body: Record<string, unknown> = {}
  if (payload.question !== undefined) body.question = payload.question
  if (payload.answer !== undefined) body.answer = payload.answer
  if (payload.meta !== undefined) body.meta = payload.meta
  if (payload.folderIds !== undefined) body.folder_ids = payload.folderIds

  const data = await apiFetch<ApiFlashcard>(`/flashcards/${cardId}`, { method: 'PUT', body })
  return mapFlashcard(data)
}

export const createFlashcard = async (payload: {
  notebookId: string
  question: string
  answer: string
  meta?: Record<string, unknown> | null
  folderIds?: string[]
}): Promise<Flashcard> => {
  const body: Record<string, unknown> = {
    notebook_id: payload.notebookId,
    question: payload.question,
    answer: payload.answer,
  }
  if (payload.meta !== undefined) body.meta = payload.meta
  if (payload.folderIds !== undefined) body.folder_ids = payload.folderIds

  const data = await apiFetch<ApiFlashcard>('/flashcards', { method: 'POST', body })
  return mapFlashcard(data)
}

export const generateFlashcardsForNotebook = async (
  notebookId: string,
  payload?: {
    attachmentIds?: string[]
    count?: number
    focus?: string
    folderName?: string
    modelKey?: string
    folderId?: string
  },
): Promise<{ folder: FlashcardFolder; flashcards: Flashcard[] }> => {
  const body: Record<string, unknown> = {}
  if (payload?.attachmentIds) body.attachment_ids = payload.attachmentIds
  if (payload?.count) body.count = payload.count
  if (payload?.focus) body.focus = payload.focus
  if (payload?.folderName) body.folder_name = payload.folderName
  if (payload?.folderId) body.folder_id = payload.folderId
  if (payload?.modelKey) body.model_key = payload.modelKey

  const data = await apiFetch<{ folder: ApiFlashcardFolder; flashcards: ApiFlashcard[] }>(
    `/notebooks/${notebookId}/flashcards/generate`,
    { method: 'POST', body },
  )

  return {
    folder: mapFlashcardFolder(data.folder),
    flashcards: (data.flashcards ?? []).map(mapFlashcard),
  }
}
